compiler/rustc_infer/src/infer/opaque_types.rs

```diff
@@ -490,6 +499,19 @@ where
                 substs.as_generator().yield_ty().visit_with(self);
                 substs.as_generator().resume_ty().visit_with(self);
             }
+            ty::Opaque(def_id, substs) => {
+                // Don't go into the same opaque type again.
+                if ty != self.skip {
+                    let skip = std::mem::replace(&mut self.skip, ty);
+                    // Only lifetimes in the bounds are relevant.
+                    // This is sound because we only permit lifetimes from the bounds to appear in the hidden type.
+                    let predicates = self.tcx.bound_explicit_item_bounds(def_id);
+                    for (predicate, _) in predicates.0 {
+                        predicates.rebind(*predicate).subst(self.tcx, substs).visit_with(self);
+                    }
```

Comment:
One problem with this is that it also ignores lifetimes substituted in place of a parameter type, and because parameter types are implicitly captured, they shouldn't be ignored. We can tweak the original test:
```diff
- impl<'a> Callable for &'a () {
+ impl<T> Callable for T {
    type Output = impl Sized;
    fn call() -> Self::Output {}
}
```
Now the test shouldn't pass, but it does.

Side-note: I really hate the rule of "explicit for lifetime, implicit for type". I can only imagine the amount of confusion this behavior brings to the average language user.

Reply:
We could make everything explicit and explore how it goes and later relax it if we want to.

Reply:
> One problem with this is that it also ignores lifetimes substituted in place of a parameter type, and because parameter types are implicitly captured, they shouldn't be ignored.

damn ^^ my first impl actually caught that. That's what I get for trying to write "smarter" code.

Reply:
> We could make everything explicit and explore how it goes and later relax it if we want to.

I like that.

Unfortunately we're stuck with this behavior for RPIT. I'm curious to see the size of impact for such change to RPIT.

Reply:
> Unfortunately we're stuck with this behavior for RPIT. I'm curious to see the size of impact for such change to RPIT.

probably massive. The closest we could get is to add something to the bounds that makes them opt out of capturing generic types and then have to opt in for the ones you want captured individually